Mathlib/Data/List/Defs.lean

```diff
@@ -749,4 +749,87 @@ Example: if `f : ℕ → list ℕ → β`, `List.mapWithComplement f [1, 2, 3]`
 def mapWithComplement {α β} (f : α → List α → β) : List α → List β :=
   mapWithPrefixSuffix fun pref a suff => f a (pref ++ suff)
 
+/--
+`List.choose` filters list `l` of type `List α` with `chooser` which is of type `α → Option β`,
+yet `β` is ignored.
+It keeps the elements where `chooser x` is `some _` and discards the ones that are `none`
+
+Example:
+```
+def oneTwo? : Nat → Option Nat
+  | 1 => some 1
+  | 2 => some 2
+  | _ => none
+
+[1, 2, 3, 5, 1, 4, 2].choose oneTwo? = [1, 2, 1, 2]
+```
+-/
+def choose (chooser : α → Option β) (l : List α) : List α :=
+  l.filter (Option.isSome ∘ chooser)
+
+/--
+Auxiliary function for `List.compareWith`
+-/
+def compareWithAux (comparer : α → α → Int) : List α → List α → Int
```

Comment:
```suggestion
def compareWithAux (comparer : α → α → Ordering) : List α → List α → Ordering
```

We have a nice type for this!

Reply:
This was once a typeclass, is it still in lean 4?

Reply:
`Ord` is still a typeclass in lean 4. I just checked

Reply:
What I'm saying is that there are two reasonable ways to write `compareWith`:
1. `compareWith₁ (comparer : α → α → Ordering) : List α → List α → Ordering`, and
2. `compareWith₂ [Ord α] : List α → List α → Ordering`.

By "it's easy to reach to the type class" I meant that you can easily write `l.compareWith₂` as `l.compareWith₁ compare` (but not the other way around).

Reply:
Ah I see.

Reply:
No the other way around.  Please do `compareWith₂`.

Reply:
And still use recursion I assume?